File: library/android/src/main/java/it/iotinga/blelibrary/BleScanCallback.java
package it.iotinga.blelibrary;

import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.bluetooth.le.ScanSettings;
import android.os.ParcelUuid;
import android.util.Log;

import androidx.annotation.RequiresPermission;

import java.util.List;

public class BleScanCallback extends ScanCallback {
  private final EventEmitter eventEmitter;

  BleScanCallback(EventEmitter eventEmitter) {
    super();
    this.eventEmitter = eventEmitter;
  }

  @Override
  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  public void onScanResult(int callbackType, ScanResult result) {
    Log.i(Constants.LOG_TAG, String.format("got scan result: %s (callback type: %d)", result, callbackType));

    boolean available = callbackType == ScanSettings.CALLBACK_TYPE_FIRST_MATCH
      || callbackType == ScanSettings.CALLBACK_TYPE_ALL_MATCHES;

    RNEventScanResult event = new RNEventScanResult();
    event.add(result, available);

    eventEmitter.emit(event);
  }

  @Override
  public void onScanFailed(int errorCode) {
    Log.e(Constants.LOG_TAG, "SCAN FAILED, error = " + errorCode);

    eventEmitter.emit(new RNEventError(BleError.ERROR_SCAN, "scan error, native error code " + errorCode));
  }

  @Override
  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  public void onBatchScanResults(List<ScanResult> results) {
    Log.i(Constants.LOG_TAG, "got batch result: " + results);

    RNEventScanResult event = new RNEventScanResult();
    for (ScanResult result : results) {
      event.add(result, true);
    }

    eventEmitter.emit(event);
  }
}

File: android/src/main/java/it/iotinga/blelibrary/BleScanCallback.kt
package it.iotinga.blelibrary

import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresPermission

class BleScanCallback internal constructor(private val sendEvent: (event: Event, payload: Map<String, Any>) -> Unit) :
  ScanCallback() {
  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  override fun onScanResult(callbackType: Int, result: ScanResult) {
    Log.i(
      LOG_TAG,
      String.format("got scan result: %s (callback type: %d)", result, callbackType)
    )

    val available = callbackType == ScanSettings.CALLBACK_TYPE_FIRST_MATCH
      || callbackType == ScanSettings.CALLBACK_TYPE_ALL_MATCHES

    sendScanResult(listOf(result), available)
  }

  override fun onScanFailed(errorCode: Int) {
    Log.e(LOG_TAG, "SCAN FAILED, error = $errorCode")

    sendEvent(
      Event.ERROR, mapOf(
        "code" to BleError.ERROR_SCAN.name,
        "message" to "scan error, native error code $errorCode",
      )
    )
  }

  @RequiresPermission(value = "android.permission.BLUETOOTH_CONNECT")
  override fun onBatchScanResults(results: MutableList<ScanResult?>) {
    Log.i(LOG_TAG, "got batch result: $results")

    sendScanResult(results, true)
  }

  private fun sendScanResult(items: List<ScanResult?>, isAvailable: Boolean) {
    sendEvent(
      Event.SCAN_RESULT, mapOf(
      "devices" to items.filterNotNull().map { item ->
        mapOf(
          "rssi" to item.rssi,
          "isAvailable" to isAvailable,
          "id" to item.device.address,
          "isConnectable" to if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) item.isConnectable else true,
          "txPower" to if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) item.txPower else 0,
          "name" to (item.scanRecord?.deviceName ?: "")
        )
      }
    ))
  }
}

